### PR TITLE
WindowAggExec output ordering is fixed

### LINKED ISSUE #4438: Fix output_ordering of WindowAggExec
**Describe the bug**

`output_ordering` method of `WindowAggExec` gives output ordering of its input, as in the [link](https://github.com/apache/arrow-datafusion/blob/fdc83e8524df30ac5d0ae097572b7c48dc686ba9/datafusion/core/src/physical_plan/windows/window_agg_exec.rs#L125). Although conceptually it is the correct behavior, `SortExec` is added before the `WindowAggExec` during `BasicEnforcement` pass of the `PhysicalOptimizerRule` (relevant [section](https://github.com/apache/arrow-datafusion/blob/fdc83e8524df30ac5d0ae097572b7c48dc686ba9/datafusion/core/src/physical_optimizer/enforcement.rs#L98)). Hence if we inquire the `output_ordering` of the `WindowAggExec` before `SortExec` is added, we will get wrong orderings.

**To Reproduce**

N.A

**Expected behavior**

I expect to get true output ordering for `WindowAggExec` during physical plan construction (before optimizations are applied)

**Additional context**

N.A

### PR DESCRIPTION
**Pull request by mustafasrepo**
# Which issue does this PR close?

Closes #4438.

# Rationale for this change

We can't rely on the current `output_ordering` result unless we pass from the `BasicEnforcement` optimization pass. Hence  `output_ordering` of the `WindowAggExec` would give the wrong result during physical plan [planner](https://github.com/apache/arrow-datafusion/blob/fdc83e8524df30ac5d0ae097572b7c48dc686ba9/datafusion/core/src/physical_plan/planner.rs#L463)

# What changes are included in this PR?
Since `WindowAggExec` maintains input order and requires specific ordering, output ordering would be required input ordering. I have changed code to reflect that (By this way we can rely on `output_ordering` information of `WindowAggExec` even if there is currently no `SortExec` before it in the physical plan)

# Are these changes tested?

N.A

# Are there any user-facing changes?

N.A

### CODE REVIEW

File: datafusion/core/src/physical_plan/windows/window_agg_exec.rs [hunk 1]
```diff
@@ -122,7 +122,9 @@ impl ExecutionPlan for WindowAggExec {
     }
 
     fn output_ordering(&self) -> Option<&[PhysicalSortExpr]> {
-        self.input.output_ordering()
+        // This executor maintains input order, and has required input_ordering filled
+        // hence output_ordering would be `required_input_ordering`
+        self.required_input_ordering()[0]
```
Comment (alamb) [inline datafusion/core/src/physical_plan/windows/window_agg_exec.rs:127]: I suggest changing this to `self.input.output_ordering()` to make the intent clearer

I realize that `required_input_ordering()` should produce the same result

**Review by alamb — APPROVED** [review]
LGTM -- I wonder if there is some way to test this to avoid regressions

Note I did something similar for `CoalesceBatchesExec`  which might provide some inspiration https://github.com/apache/arrow-datafusion/pull/4332

File: datafusion/core/src/physical_plan/windows/window_agg_exec.rs (hunk 1 above)
Comment (ozankabak) [inline datafusion/core/src/physical_plan/windows/window_agg_exec.rs:127]: In order to be a little easier to read, we can use `self.sort_keys.as_deref()` here. This would avoid the construction of a single-element vector and then taking the first element of it. The downside of doing this would be replicating this line from `required_input_ordering`, but I think the added clarity is worth it here. Thoughts, @alamb?

Comment (alamb) [reply datafusion/core/src/physical_plan/windows/window_agg_exec.rs:127]: that would also be good

**Comment by mingmwang** [comment]
No, I think this change is not correct!!!  WindowAggExec should always keep the input ordering, and the input ordering does not always equal with the required input ordering. For example, input ordering could be `order by ['a', 'b', 'c']`, the required input ordering is `order by ['a', 'b']`,  `order by ['a', 'b', 'c']` can satisfy `order by ['a', 'b']`, in this case, the output ordering of the WindowAggExec itself is still `order by ['a', 'b', 'c']`.

**Comment by mingmwang** [comment]
@mustafasrepo Could you please explain more specifically why you can not rely on the current `output_ordering` result?
Do you have your own physical optimizer rules?

**Comment by mustafasrepo** [comment]
> #4332
@mingmwang Thanks for your feedback. I saw this problem while working on a PR that removes unnecessary `SortExec` before `WindowAggExec` to improve efficiency. Consider the query below 

```sql
SELECT
      c9,
      SUM(c9) OVER(ORDER BY c9) as sum1,
      SUM(c9) OVER(ORDER BY c9, c8) as sum2
      FROM aggregate_test_100
```

Its physical plan is as follows 

```sql
+---------------+-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| plan_type     | plan                                                                                                                                                                                                                                                |
+---------------+-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| logical_plan  | Projection: aggregate_test_100.c9, SUM(aggregate_test_100.c9) ORDER BY [aggregate_test_100.c9 ASC NULLS LAST] AS sum1, SUM(aggregate_test_100.c9) ORDER BY [aggregate_test_100.c9 ASC NULLS LAST, aggregate_test_100.c8 ASC NULLS LAST] AS sum2     |
|               |   WindowAggr: windowExpr=[[SUM(aggregate_test_100.c9) ORDER BY [aggregate_test_100.c9 ASC NULLS LAST]]]                                                                                                                                             |
|               |     WindowAggr: windowExpr=[[SUM(aggregate_test_100.c9) ORDER BY [aggregate_test_100.c9 ASC NULLS LAST, aggregate_test_100.c8 ASC NULLS LAST]]]                                                                                                     |
|               |       TableScan: aggregate_test_100 projection=[c8, c9]                                                                                                                                                                                             |
| physical_plan | ProjectionExec: expr=[c9@3 as c9, SUM(aggregate_test_100.c9) ORDER BY [aggregate_test_100.c9 ASC NULLS LAST]@0 as sum1, SUM(aggregate_test_100.c9) ORDER BY [aggregate_test_100.c9 ASC NULLS LAST, aggregate_test_100.c8 ASC NULLS LAST]@1 as sum2] |
|               |   WindowAggExec: wdw=[SUM(aggregate_test_100.c9): Ok(Field { name: "SUM(aggregate_test_100.c9)", data_type: UInt64, nullable: true, dict_id: 0, dict_is_ordered: false, metadata: None })]                                                          |
|               |     SortExec: [c9@2 ASC NULLS LAST]                                                                                                                                                                                                                 |
|               |       WindowAggExec: wdw=[SUM(aggregate_test_100.c9): Ok(Field { name: "SUM(aggregate_test_100.c9)", data_type: UInt64, nullable: true, dict_id: 0, dict_is_ordered: false, metadata: None })]                                                      |
|               |         SortExec: [c9@1 ASC NULLS LAST,c8@0 ASC NULLS LAST]                                                                                                                                                                                         |
|               |           CsvExec: files=[Users/akurmustafa/projects/synnada/arrow-datafusion-synnada/testing/data/csv/aggregate_test_100.csv], has_header=true, limit=None, projection=[c8, c9]                                                                    |
|               |                                                                                                                                                                                                                                                     |
+---------------+-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
```

Second (from the bottom) `SortExec` can be removed from the physical plan safely. To remove unnecessary `SortExec` from the physical plan, I compare the required sorting of `WindowAggExec` with its `input_exec.output_ordering()` at [here](https://github.com/apache/arrow-datafusion/blob/fdc83e8524df30ac5d0ae097572b7c48dc686ba9/datafusion/core/src/physical_plan/planner.rs#L518). Second `WindowAggExec` requires input_ordering `c9@2 ASC NULLS LAST`. Its input `WindowAggExec` should have `output_ordering` , `c9@1 ASC NULLS LAST,c8@0 ASC NULLS LAST`. However, `output_ordering` of the first `WindowAggExec` returns `None`. The reason is that since we did not put `SortExec` before it yet (It is is done at optimization pass), its input is `CsvExec` and its `ouput_ordering` is `None`. 
In summary, unless we pass from `BasicEnforcement` optimization pass, the output ordering information of `WindowAggExec` would give output ordering of Executor before `SortExec` in the final physical plan.


> @mustafasrepo Could you please explain more specifically why you can not rely on the current `output_ordering` result? Do you have your own physical optimizer rules?

**Comment by mustafasrepo** [comment]
> @mustafasrepo Could you please explain more specifically why you can not rely on the current `output_ordering` result? Do you have your own physical optimizer rules?

To replicate my problem if we put the statement `println!("input exec output ordering:{:?}", input_exec.output_ordering()); `[here](https://github.com/apache/arrow-datafusion/blob/fdc83e8524df30ac5d0ae097572b7c48dc686ba9/datafusion/core/src/physical_plan/planner.rs#L518)
For above query, we receive output ordering `input exec output ordering:None`, for both `WindowAggExec`s

However, for the second  `WindowAggExec` we should receive `input exec output ordering:Some([PhysicalSortExpr { expr: Column { name: "c9", index: 1 }, options: SortOptions { descending: false, nulls_first: false } }, PhysicalSortExpr { expr: Column { name: "c8", index: 0 }, options: SortOptions { descending: false, nulls_first: false } }])`.

**Comment by mingmwang** [comment]
@mustafasrepo `BasicEnforcement` should do the optimization and avoid the unnecessary SortExec for you. I will write a test in DataFusion late today and try to reproduce the issue.

**Comment by mingmwang** [comment]
@mustafasrepo `BasicEnforcement` should do the optimization and avoid the unnecessary SortExec for you. I will write a test in DataFusion late today and try to reproduce the issue.

**Comment by ozankabak** [comment]
@mingmwang, even though @mustafasrepo's example shows how the issue can manifest in a certain context, the core issue is that `input_exec.output_ordering()` can return a result that is inconsistent with `self. required_input_ordering()`. In his specific case, the former returns `None`, which is clearly wrong -- the requirement would have been violated had this been the case.

This seems to happen because as of PR #4122, we insert operators like`SortExec` *after* physical planning (i.e. during the `BasicEnforcement` optimization step). Therefore, calls like `input_exec.output_ordering()` return such inconsistent/unreliable results during physical planning, since the "real" `input_exec` is not there yet!

Until we figure out a general fix to this, we can select the "finer" ordering between `required_input_ordering()` and `input.output_ordering()` -- this will result in a less wrong result in the meantime.

**Comment by mingmwang** [comment]
> @mingmwang, even though @mustafasrepo's example shows how the issue can manifest in a certain context, the core issue is that `self.input.output_ordering()` can return a result that is inconsistent with `self. required_input_ordering()`. In his specific case, the former returns `None`, which is clearly wrong -- the requirement would have been violated had this been the case.
> 
> This seems to happen because as of PR #4122, we insert operators like`SortExec` _after_ physical planning (i.e. during the `BasicEnforcement` optimization step). Therefore, calls like `input_exec.output_ordering()` return such inconsistent/unreliable results during physical planning, since the "real" `input_exec` is not there yet!
> 
> Until we figure out a general fix to this, we can select the "finer" ordering between `required_input_ordering()` and `input.output_ordering()` -- this will result in a less wrong result in the meantime.

There is a general fix.  The root cause is not because `output ordering` return None, but when the `WindowAggExec` try to derive the ordering from input, the column index must be handle properly. The same bug exists within output partition.



I will raise a PR to fix the bug.

**Comment by mustafasrepo** [comment]
> No, I think this change is not correct!!! WindowAggExec should always keep the input ordering, and the input ordering does not always equal with the required input ordering. For example, input ordering could be `order by ['a', 'b', 'c']`, the required input ordering is `order by ['a', 'b']`, `order by ['a', 'b', 'c']` can satisfy `order by ['a', 'b']`, in this case, the output ordering of the WindowAggExec itself is still `order by ['a', 'b', 'c']`.

We have updated this PR to address this concern. Thanks for pointing it out @mingmwang .

**Comment by metesynnada** [comment]
Let's consider more general use cases of `output_ordering()`. In `DefaultPhysicalPlanner.create_initial_plan()`, we should be able to depend on `output_ordering()` and `output_partitioning()` regardless of the optimization step.  Let’s think we have `AExec` and `BExec`.

```
AExec

  ⬇️

BExec
```

where 

```rust
impl ExecutionPlan for AExec {
    fn output_ordering(&self) -> Option<&[PhysicalSortExpr]> {
        self.input.output_ordering()
    }
}
```

If I call `AExec.output_ordering()` in `DefaultPhysicalPlanner`, it will gives `BExec`’s output order. I can make decisions depending on it in planner phase. To make this plan “runnable”, we should not add new operators to the between `AExec` and `BExec` outside of the planner. If `AExec` requires an input ordering, this should be decided inside the planner and add to the operator chain to make `output_ordering()` API work. Currently, these API’s are broken since we did not add a `SortExec` inside the planner.

**Comment by ozankabak** [comment]
Great discussion. I think @metesynnada highlighted the core issue nicely. I don't yet understand the general fix @mingmwang has in mind, but if it works, this PR will become unnecessary.

If it doesn't solve the problem @metesynnada explained, we can use this fix until a final, general fix is available.

**Comment by mingmwang** [comment]
@metesynnada 
The rule `BasicEnforcement` is part of the `DefaultPhysicalPlanner`, that's why the rule is an enforcement rule(A common rule in many other DB Engines).

**Comment by mingmwang** [comment]
I just submit the fix PR, please help to take a look.

https://github.com/apache/arrow-datafusion/pull/4455

**Comment by mingmwang** [comment]
The `BasicEnforcement` can choose to satisfy given ordering/partitioning requirements in different ways. For example a requirement ordering `['a'] `can be satisfied by sort by `['a']`, sort by` ['a', 'b']`, order by `['a', 'b', 'c']`, sort by `['a', random]` ......
In future, we will add support for Range Partitioning, if the optimizer detects the column a is a skewed column,   it can effectively add  sort by `[a + random()] `to avoid data skew in the Range Partitioning.

**Comment by ozankabak** [comment]
Thanks! Let's conclude the discussion on your PR first, and then we can easily decide whether to go forward with this or not.

**Comment by ozankabak** [comment]
The need for this PR was obviated by PR ##4455 thanks to @mingmwang. We will be closing this shortly and focus our efforts on creating a rule that will swap the default `WindowAggExec` with more-efficient, pipeline-able version when appropriate.
